### ISSUE TITLE
AOT compiler emits redundant StoreInstanceField instructions

### ISSUE THREAD

**Issue by ds84182** [dart-lang/sdk#49657]
Godbolt Link: https://dart.godbolt.org/z/W7cr6afjv

```
*** BEGIN CFG
After AllocateRegisters
==== file://<source>_::_main (RegularFunction)
  0: B0[graph]:0 {
      v292 <- UnboxedConstant(#6.0 double) T{_Double}
      v293 <- UnboxedConstant(#8.0 double) T{_Double}
      v294 <- UnboxedConstant(#10.0 double) T{_Double}
      v295 <- UnboxedConstant(#11.0 double) T{_Double}
      v296 <- UnboxedConstant(#13.0 double) T{_Double}
      v297 <- UnboxedConstant(#15.0 double) T{_Double}
}
  2: B1[function entry]:2
  4:     v9 <- AllocateObject:16(cls=Vec3Mut, <not-aliased>) T{Vec3Mut}
  5:     ParallelMove rax <- rax, xmm0 <- C
  6:     StoreInstanceField(v9 . _x@13416190 = v292, NoStoreBarrier)
  7:     ParallelMove xmm0 <- C
  8:     StoreInstanceField(v9 . _y@13416190 = v293, NoStoreBarrier)
  9:     ParallelMove xmm0 <- C
 10:     StoreInstanceField(v9 . _z@13416190 = v294, NoStoreBarrier)
 11:     ParallelMove xmm0 <- C
 12:     StoreInstanceField(v9 . _x@13416190 = v295, NoStoreBarrier)
 13:     ParallelMove xmm0 <- C
 14:     StoreInstanceField(v9 . _y@13416190 = v296, NoStoreBarrier)
 15:     ParallelMove xmm0 <- C
 16:     StoreInstanceField(v9 . _z@13416190 = v297, NoStoreBarrier)
 17:     ParallelMove rax <- rax
 18:     Return:30(v9)
*** END CFG
```

Looks like StoreInstanceField isn't being elided, which keeps a set of intermediate constants alive.

cc @mraleph

**Comment by mraleph** [comment]
I did a quick look. This is because of interfering `AllocateObject` instructions which are later eliminated by the allocation sinking pass. `AllocateObject` instruction is marked as throwing and throwing is unconditionally treated as "load everything". In reality we could have a more precise condition here - throwing is load everything only if catching code can see the destination of the write.

**Comment by aam** [comment]
Could another factor be that [initializing stores](https://github.com/dart-lang/sdk/blob/main/runtime/vm/compiler/backend/redundancy_elimination.cc#L3047) (coming from constructors) are not going to be eliminated by dead store elimination?

**Comment by aam** [comment]
as @rmacnak-google noted, potentially, if there is no longer a need to have special handling of shared box initialization after https://dart-review.googlesource.com/c/sdk/+/255250, we could get rid of `CanEliminateStore` in redundancy elimination

**Comment by mraleph** [comment]
@aam yeah, and we should not care about initialising stores in AOT anyway.

**Comment by aam** [comment]
@mraleph , ah, good point. Might be as simple as https://dart-review.googlesource.com/c/sdk/+/256425 then

**Comment by mraleph** [comment]
@aam was this marked fixed by accident? the thing with allocation instructions interfering with DSE (https://github.com/dart-lang/sdk/issues/49657#issuecomment-1214976047) is still not fixed so redundant stores are still there.

**Comment by aam** [comment]
No, I thought I checked original sample, but it looks I only verified my simplified version(included as a test in the cl) which was not affected by `AllocateObject` instructions.

**Comment by aam** [comment]
> throwing is load everything only if catching code can see the destination of the write.

Wanted to confirm: try-catch block in the function disables AllocationSinking, removal of allocation instructions, so there are no redundant stores to remove(`baz_try_catch` below).

Given 
```dart
class Foo {
  int x = -1;

  toString() => "Foo x: $x";
}

class Bar {}

var arr = <dynamic>[];

main() {
  main_nocatch();
  main_catch();
}

main_nocatch() {
  {
    var foo = new Foo();
    foo = baz_param(foo);
    print(foo);
    print(arr[0]);
  }

  {
    var foo = baz_escape();
    print(foo);
    print(arr[1]);
  }

  {
    var foo = baz_noescape();
    print(foo);
  }

  {
    var foo = baz_try_catch();
    print(foo);
  }
}

main_catch() {
  {
    var foo = new Foo();
    try {
      foo = baz_param(foo);
    } catch (_) {}
    print(foo);
    print(arr[0]);
  }

  {
    try {
      var foo = baz_escape();
      print(foo);
    } catch(_) {}
    print(arr[1]);
  }
}

// 12, 13, 14, 15
baz_param(var foo) {
  foo.x = 12;
  new Bar();
  foo.x = 13;
  new Bar();
  arr.add(foo);

  foo.x = 14;
  new Bar();
  foo.x = 15;
  return foo;
}

// 14, 15
baz_escape() {
  var foo = Foo();
  //...
  foo.x = 12;
  new Bar();
  foo.x = 13;
  new Bar();
  foo.x = 14;
  arr.add(foo);
  new Bar();
  foo.x = 15;
  return foo;
}

// 12, 13, 14
baz_noescape() {
  var foo = Foo();
  foo.x = 12;
  new Bar();
  foo.x = 13;
  new Bar();
  foo.x = 14;
  return foo;
}

// 12, 13, 14
baz_try_catch() {
  var foo = Foo();
  try {
    foo.x = 12;
    new Bar();
    foo.x = 13;
    new Bar();
    foo.x = 14;
  } catch (_) {

  }
  return foo;
}
```
all of the functions except `baz_try_catch` current have redundant stores. Did you have particular ones in mind that we could start with?
Should we look at `baz_noescape` where all stores are local(`foo` is local variable), so they won't escape the method context even if exception is thrown?

Another thought is that we could simply repeat DSE pass after AllocationSinking_Sink(that removes dead `AllocateObject` instructions), which automagically solves all cases of redundant stores(except `baz_try_catch` where stores remain relevant since AllocateObject instructions remain).

**Comment by aam** [comment]
> There is a bit of complexity here with lazy deoptimization unfortunately, `AllocateObject` can lazily deoptimise so an object can become visible through that,

Sorry, I don't get how potentially-throwing `AllocaleObject` can change visibility of local "not-aliased" var.
In the example, code might get optimized for `f.f is int`, so generated optimized code can only produce `x.value == 2` through second `return`, can't throw because `AllocateObject` instruction for `Bar` is gone.
If deopt hits for `f.f is! int`, then unoptimized code can see either `x.value == 1` through first `return` or `x.value == 2` through second `return`. It can also throw on `Bar()` allocation since `AllocateObject` instruction is reinstated in unoptimized code.
But `Bar()` `AllocateObject` doesn't introduce different escape path for local `x`, doesn't change it's visibility by virtue that it may throw, so unless `f.f is! int`, caller won't see `x.value == 1`.

**Comment by mraleph** [comment]
@aam it's not related to throwing, it is related to (lazy) deoptimization. You are correct that throwing instructions don't expose not-aliased allocations (assuming no catch in the function - because most of these optimizations are disabled by catch anyway as discussed). Deoptimization (be it lazy or eager, FWIW) can expose an object that otherwise looks unexposed, like example above shows because optimizing compiler looks at a different version of the graph. 

Optimizing compiler might reduce the code above to:

```dart
Foo foo(Flag f) {
  final x = Foo();
  x.value = 1;
  Bar(); 
  // if folded away and removed  
  x.value = 2;
  return x;
}
```

In this example `x.value == 1` can't be observed, but if there is a deoptimization at `Bar()` the execution will then continue in a full function - and in full function `x.value == 1` can potentially be observed.

The good model here is to treat deoptimization `env` as a side-exit which exposes the current state of all values mentioned in the `env`.

**Comment by aam** [comment]
> The good model here is to treat deoptimization env as a side-exit which exposes the current state of all values mentioned in the env.

Interesting, I see.
But `AllocationInstr` [can deopt](https://github.com/dart-lang/sdk/blob/main/runtime/vm/compiler/backend/il.h#L6767), therefore if we are to generically assume that for any instruction that can deopt, everything in it's environment can escape, that would include `x`. Therefore, we won't be able to get rid of `x.value = 1` even if there is nothing other than that `Bar()` allocation between `x.value = 1` and `x.value = 2`.
Should we apply this "instruction can expose its environment" only to instruction that not [MayThrow](https://github.com/dart-lang/sdk/blob/main/runtime/vm/compiler/backend/il.h#L1301)? Or even that can't apply to all instructions?

**Comment by mraleph** [comment]
For simplicity we could implement improvements to DSE in AOT only (again). In AOT we could just ignore deoptimizations altogether.
